pep-9999.rst

```diff
@@ -0,0 +1,217 @@
+    url                 str          URL to repository
+    ==================  ===========  =============================================
+
+
+The auth.toml file
+''''''''''''''''''
+
+Credentials are stored in the ``auth.toml`` file, under the ``[authentication]`` section. There is a correlation between the repository name and the authentication section's name::
+
+    [authentication]
+
+    [authentication.pypi]
+    username = "__token__"
+    password = "pypi-AgEIcHlwaS5vcmcCJDJjMmI5NDJmLWIzNzQtNDc3ZC04ZmRkLTM4NDEzY2M1ZTU5MAACJXsicGVybWlzc2lvbnMiOiAidXNlciIsICJ2ZXJzaW9uIjogMX0AAAYgV91r8iEUUz7Q5xbsfi_KGiO3wrJR2_IOOpnFyLo62bk"
+
+    [authentication.myrepo1]
+    username = "foo"
+    password = "YmFyCg=="
+
+
+Instead of specifying the credentials in this file, Keyring can be used::
+
+    keyring set https://pypi.org/simple/ __token__
+    keyring set <repository URL> <private-repository username>
+
+
+A repository's credentials has the following attributes::
+
+    ==================  ===========  ===============================================
+    Attribute           Type         Description
+    ==================  ===========  ===============================================
+    cert                str          Path to certificate authority bundle (*.pem)
+    cert-client         str          Path to client certificate (*.pem)
+    password            str          The password for the repository, base64-encoded. When token is used, enter it here in clear-text.
+    proxy_password      str          The password for the proxy server, base64-encoded.
+    proxy_username      str          The username for the proxy server
+    username            str          The username for the repository. When token is used, set this to ``__token__``.
+    ==================  ===========  ===============================================
+
+
+Package discovery
+-----------------
+
+The PyPI repository (``repositories.pypi``) does not have to be specified in the ``config.toml``. If not specified, it will be used as the first package index by default. However, by explicitly defining it, the repository attributes can be used and it can be ordered differently.
+
+The order of the repository definitions in ``config.toml`` shall be respected when searching for packages. This allows custom package indexes to take precendence over e.g. a default PyPI repository and the use of package shadowing.
+
+
+Configuration file discovery
+----------------------------
+
+The locations of the configuration files vary slightly across platforms. You may have per-user, per-virtualenv or global (shared amongst all users) configuration:
+
+If multiple configuration files are found, then they are combined (merge and override) in the following order:
+
+1. The global file is read
+2. The per-user file is read
+3. The virtualenv-specific file is read
+
+Each file read overrides any values read from previous files.
```

Comment:
TOML is implicitly nested dictionaries, I believe, so you may want to define this rule a bit more robustly. A top-level `.update` is going to lose some information, while deep-update may also be complex due to missing keys and default values. There's also the matter of adding more indexes and preserving order, which dictionaries only kind-of do (and you certainly shouldn't assume it in this spec).

Reply:
Yes, this needs to be defined more explicitly. My thought was a deep merge of the dict.

What you're saying about the ordering... makes me finally want to drop the idea of retaining an order at all. At least not an order the way I've described it in this draft now.

Perhaps a new `order_id` attribute could be introduced, which can be an `int`, if order is important to support.

Reply:
A `priority` attribute would be fine, and all you have to do is say that it's an int, should be sorted descending, and defaults to zero. That definition is easiest to implement and reason about.

However, I don't think any tools (certainly not pip) are able to apply an ordering to their indexes today, so they'll likely push back on it. So perhaps specify it like:

> Tools SHOULD respect `priority` by acting as if all lower priority repositories were not specified when a higher priority repository satisfies the task. Tools that choose not to respect `priority` MUST display a warning to users when a non-zero priority value is being ignored. It is acceptable for tools to be less efficient in order to handle prioritized repositories, for example, by not performing queries in parallel.

Reply:
This is great. I would like to take your statement as-is into the draft.

At this point I think I should put your name on the draft as well. Would you mind if I did that?
I mean, so to give credit where credit is due 😄

Reply:
Go for it, and don't worry about putting my name on it. You can say I've helped review it already, but plenty of people will end up contributing as much as I have.

Reply:
Ok, I don't have any experience with writing a PEP, but if you keep this up it seems you should be part of the "Authors" in the metadata. :)